### PR TITLE
feat(Storybook): Hot reloading :fire:

### PR DESCRIPTION
**Pull request by micieslak**
## What does the PR do

Closes: https://github.com/status-im/status-desktop/issues/7975

This PR implements to variants of hot reloading for Storybook:

1. Page-wise hot reloading.
    When the source is changed, the currently opened page is reloaded. The state of the page is gone. It doesn't require any special treatment when it comes to page implementation. The mechanism is fully transparent.

[Kazam_screencast_00068.webm](https://user-images.githubusercontent.com/20650004/196904022-024f7024-ae5c-4a2e-acb7-d0767df5d9cc.webm)

2. In-page hot reloading.
In comparison to generic page-wise hot reloading, this technique requires using dedicated component within a storybook page but provides greater flexibility e.g. preserving component state across source reloads. For details check out the example page `StatusCommunityCard`. When in-page variant is used, page-wise hot reloading should be disabled to avoid reloading the whole page and loosing component's state (checkbox in the left column). When page reload is needed, e.g. after changing controls for manipulating component's properties, reload can be enforced by `Reload now`. Additionally this variant of hot reloading allows to expose code instantiating given component to the user in order to modify it on the fly.

[Kazam_screencast_00070.webm](https://user-images.githubusercontent.com/20650004/196904925-ad7f5d39-e555-45f7-a08d-3811d6cf3820.webm)

### Known limitations
- Singleton instances are not dropped and recreated when sources change. It means that hot reloading doesn't reflect changes in components like `Theme`, `Global` or `Utils`. Starting from `Qt 6.4` it should be easy to do using https://doc.qt.io/qt-6/qqmlengine.html#clearSingletons, however probably it will be necessary to change the way in which light/dark mode is handled now in Storybook.

### CODE REVIEW

File: storybook/directorieswatcher.h [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    Q_OBJECT
+public:
+    explicit DirectoriesWatcher(QObject *parent = nullptr);
+    void addPaths(const QStringList &paths);
+
+signals:
+    void changed();
+
+private:
+    QFileSystemWatcher* fsWatcher;
```
Comment (osmaczko) [inline storybook/directorieswatcher.h:18]: Why this is a pointer? There is no need to create it using new.
```suggestion
    QFileSystemWatcher fsWatcher;
```

File: storybook/src/Storybook/HotReloader.qml [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+QtObject {
+    id: root
+
+    /*required*/ property Loader loader
+    property bool enabled: false
+
+    signal reloaded
+
+    function forceReload() {
+        d.asyncBlocker.when = true
```
Comment (osmaczko) [inline storybook/src/Storybook/HotReloader.qml:15]: Why do we need to disable async before we unload the component?

File: storybook/src/Storybook/HotReloader.qml [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        d.sourceBlocker.when = true
+        CacheCleaner.clearComponentCache()
+        d.asyncBlocker.when = false
+        d.sourceBlocker.when = false
```
Comment (osmaczko) [inline storybook/src/Storybook/HotReloader.qml:19]:
```suggestion
        loader.active = false
        CacheCleaner.clearComponentCache()
        loader.active = true
```

File: storybook/src/Storybook/HotLoader.qml [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+Loader {
+    sourceComponent: hotComponent.component
+
+    property alias source: hotComponent.source
+    property alias rethrowErrors: hotComponent.rethrowErrors
+    readonly property alias errors: hotComponent.errors
+
+    HotComponentFromSource {
+        id: hotComponent
+        source: sourceTextEdit.text
```
Comment (osmaczko) [inline storybook/src/Storybook/HotLoader.qml:12]: Leftover?

**Review by osmaczko — APPROVED** [review]
Very nice!!

File: storybook/src/Storybook/HotReloader.qml (hunk 2 above)
Comment (micieslak) [reply storybook/src/Storybook/HotReloader.qml:15]: I added comment for that in code :)

File: storybook/src/Storybook/HotReloader.qml (hunk 3 above)
Comment (micieslak) [reply storybook/src/Storybook/HotReloader.qml:19]: as above, it's for a reason :)

File: storybook/src/Storybook/HotLoader.qml (hunk 4 above)
Comment (micieslak) [reply storybook/src/Storybook/HotLoader.qml:12]: yes, removed :)

File: storybook/directorieswatcher.h (hunk 1 above)
Comment (osmaczko) [reply storybook/directorieswatcher.h:18]: That's fair enough, although I am not a big fan of this solution (raw pointers in general, even in qtish environment). I prefer readability and maintainability over compilation times. Here, you don't know straight away who the owner of `QFileSystemWatcher` is.

Comment (micieslak) [reply storybook/directorieswatcher.h:18]: yeah, there are some pros/cons...

Comment (caybro) [reply storybook/directorieswatcher.h:18]: Yup, those raw pointers... :) I'd prefer either a stack var, or forward decl too

**Review by caybro — APPROVED** [review]
Nice work!

**Review by iurimatias — APPROVED** [review]

**Pull request merged**